File: backend/app/schemas/community.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


CommunityType = Literal["fan", "official", "platform"]
CommunityStatus = Literal["pending", "approved", "rejected", "hidden"]
CommunityMemberRole = Literal["owner", "moderator", "member"]
CommunityMemberStatus = Literal["active", "banned", "left"]
CommunityActionStatus = Literal["pending", "reviewed", "dismissed", "action_taken"]


class CommunityCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=160)
    description: str | None = None
    community_type: CommunityType = "fan"
    entity_id: int | None = None
    image_url: str | None = Field(default=None, max_length=1000)
    banner_url: str | None = Field(default=None, max_length=1000)


class CommunityUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=160)
    description: str | None = None
    entity_id: int | None = None
    image_url: str | None = Field(default=None, max_length=1000)
    banner_url: str | None = Field(default=None, max_length=1000)


class CommunityListItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    slug: str
    description: str | None
    community_type: CommunityType
    entity_id: int | None
    owner_user_id: int
    status: CommunityStatus
    image_url: str | None
    banner_url: str | None
    member_count: int
    created_at: datetime
    updated_at: datetime


class CommunityMemberRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    community_id: int
    user_id: int
    role: CommunityMemberRole
    status: CommunityMemberStatus
    joined_at: datetime


class CommunityRuleCreate(BaseModel):
    title: str = Field(..., min_length=1, max_length=160)
    description: str | None = None
    order_index: int = 0


class CommunityRuleUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=160)
    description: str | None = None
    order_index: int | None = None


class CommunityRuleRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    community_id: int
    title: str
    description: str | None
    order_index: int
    created_at: datetime
    updated_at: datetime


class CommunityRead(CommunityListItem):
    rules: list[CommunityRuleRead] = Field(default_factory=list)


class CommunityReportCreate(BaseModel):
    reason: str = Field(..., min_length=1, max_length=120)
    details: str | None = None


class CommunityReportRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    community_id: int
    reporter_user_id: int
    reason: str
    details: str | None
    status: CommunityActionStatus
    created_at: datetime
    resolved_at: datetime | None


class CommunityMergeRequestCreate(BaseModel):
    source_community_id: int
    target_community_id: int
    reason: str = Field(..., min_length=1, max_length=240)


class CommunityMergeRequestRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    source_community_id: int
    target_community_id: int
    requested_by_user_id: int
    reason: str
    status: CommunityActionStatus
    created_at: datetime
    resolved_at: datetime | None

File: frontend/types/community.ts
export type CommunityType = "fan" | "official" | "platform";
export type CommunityStatus = "pending" | "approved" | "rejected" | "hidden";
export type CommunityMemberRole = "owner" | "moderator" | "member";
export type CommunityMemberStatus = "active" | "banned" | "left";
export type CommunityActionStatus =
  | "pending"
  | "reviewed"
  | "dismissed"
  | "action_taken";

export interface CommunityListItem {
  id: number;
  name: string;
  slug: string;
  description: string | null;
  community_type: CommunityType;
  entity_id: number | null;
  owner_user_id: number;
  status: CommunityStatus;
  image_url: string | null;
  banner_url: string | null;
  member_count: number;
  created_at: string;
  updated_at: string;
}

export interface CommunityRule {
  id: number;
  community_id: number;
  title: string;
  description: string | null;
  order_index: number;
  created_at: string;
  updated_at: string;
}

export interface Community extends CommunityListItem {
  rules: CommunityRule[];
}

export interface CommunityMember {
  id: number;
  community_id: number;
  user_id: number;
  role: CommunityMemberRole;
  status: CommunityMemberStatus;
  joined_at: string;
}

export interface CommunityCreatePayload {
  name: string;
  description?: string | null;
  community_type: CommunityType;
  entity_id?: number | null;
  image_url?: string | null;
  banner_url?: string | null;
}

export interface CommunityRuleCreatePayload {
  title: string;
  description?: string | null;
  order_index?: number;
}

export interface CommunityReportPayload {
  reason: string;
  details?: string | null;
}

export interface CommunityReport {
  id: number;
  community_id: number;
  reporter_user_id: number;
  reason: string;
  details: string | null;
  status: CommunityActionStatus;
  created_at: string;
  resolved_at: string | null;
}

export interface CommunityMergeRequestPayload {
  source_community_id: number;
  target_community_id: number;
  reason: string;
}

export interface CommunityMergeRequest {
  id: number;
  source_community_id: number;
  target_community_id: number;
  requested_by_user_id: number;
  reason: string;
  status: CommunityActionStatus;
  created_at: string;
  resolved_at: string | null;
}
